working/1426-extension-types/feature-specification-implements-rep.md

```diff
@@ -0,0 +1,268 @@
+# Inline Classes can Implement the Representation Type
+
+Author: eernst@google.com
+
+Status: Draft.
+
+This is an addendum to the [inline class feature specification][],
+proposing an additional feature: An inline class `V` may have its
+representation type `R` as a superinterface. This causes `V` to be a
+subtype of `R`, and it enables invocations of members of `R` on a receiver
+of type `V`.
+
+[inline class feature specification]: https://github.com/dart-lang/language/blob/main/accepted/future-releases/inline-classes/feature-specification.md
+
+## Change Log
+
+## Summary
+
+This is a proposal to add a new feature to inline classes: An inline class `V`
+may include one or more superinterfaces in its `implements` clause which are
+supertypes `R1 .. Rk` of the ultimate representation type `R` of `V` (this
+allows `implements R` as a special case). This causes `V` to be a subtype
+of each of `Rj`, `j` in `1 .. k`, and it enables invocations of members of
+`Rj` on a receiver of type `V`.
```

Comment:
> we need to be more careful about which types are incompatible

I tend to think that there shouldn't be many new things here. In particular, it's only about member signatures, and both function types and nullable types have very few members, namely the ones of `Object`.

`NonNull(InlineType) == InlineType`, so that shouldn't create any new considerations.

If an inline type is a subtype of several nullable types and function types (and what not, insofar as this is even possible) then we'd need to assign it to one of those supertypes before we can use the properties of that specific supertype (for instance, we need to go from `InlineType implements int?` to `int?` before we can use `?.` on it.

> assignable to `int Function(int)` and `int Function(int, int)`, but not to `int Function(int, [int])`

I don't see a problem: A subtype rule about superinterfaces can confirm the connection to each of those function types, which yields the assignability to each of them, and then we're at home again. The inline type itself can't _use_ those superinterfaces to get any members unless all the supertypes agree (they say the same, or nothing).

> doesn't work, since `S` and `T` are not supertypes of `dynamic`

Agreed, that's an error. We really do insist that the representation type is a subtype of each of the non-inline superinterfaces.

> I only found problems with implementing the same interface in more than one way when I wrote something which needed a single answer.

It is possible that we can allow the same inline class to implement several different generic instantiations of the same class. However, I don't think that's going to be any better than implementing the least (most special) of those types, if it exists. There _must_ be some shared subtypes because the representation type is one such type.

Is the following example important? Why wouldn't we simply say that it is a compile-time error, and go on with our lives? ;-)

```dart
inline class A(List<num> it) implements List<num> {}
inline class B(List<int> it) implements A, List<int> {
  ... // Trying really hard, redeclare all 30-50 members that have a conflict.
}
```

Reply:
Actually, we wouldn't need to redeclare all those members as long as there is a most specific signature for each of the members (such that they have a combined member signature). This _does_ mean that we'd consider a `B` to have an `add` that accepts a `num` and a `first` that yields an `int`.

Still not sure we want this. ;-)

Reply:
ACK

> `NonNull(InlineType) == InlineType`, so that shouldn't create any new considerations.

The inline class type is not nullable because `Null` is not a subtype of it, even if the underlying object might be nullable, and the inline class itself is a subtype of `Object?` (but then, everything is).